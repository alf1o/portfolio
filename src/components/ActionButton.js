import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Message from '@material-ui/icons/Message';
import Work from '@material-ui/icons/Work';
import Code from '@material-ui/icons/Code';
import Mail from '@material-ui/icons/Mail';
import { Link } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';

const styles = theme => ({
  trigger: {
    position: 'absolute',
    backgroundColor: theme.palette.primary.main
  },
  actions: {
    position: 'absolute',
    transition: 'bottom, right 300ms ease-out',
    backgroundColor: '#fff'
  }
});

function getIcon(icon) {
  switch(icon) {
    case 'message':
      return <Message color="secondary" />;
    case 'mail':
      return <Mail color="primary" />
    case 'linkedin':
      return <Work color="primary" />
    case 'github':
      return <Code color="primary" />
    default:
      return <Message color="secondary" />;
  }
}

/**
  Render a button which will open a way of contanct.
*/
function ActionButton({ classes, pos, icon, onClick }) {
  return (
    <Button
      className={
        icon === 'message' ? classes.trigger : classes.actions
      }
      style={{ bottom: pos.bottom, right: pos.right }}
      variant="fab"
      onClick={onClick}
    >
      {getIcon(icon)}
    </Button>
  );
}

ActionButton.propTypes = {
  classes: PropTypes.object.isRequired,
  pos: PropTypes.object.isRequired,
  icon: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired
};

export { ActionButton };
export default withStyles(styles)(ActionButton);
