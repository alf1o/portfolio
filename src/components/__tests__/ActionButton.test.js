import React from 'react';
import ActionButton from '../ActionButton';
import { createShallow } from '@material-ui/core/test-utils';
import Message from '@material-ui/icons/Message';
import Work from '@material-ui/icons/Work';
import Code from '@material-ui/icons/Code';
import Mail from '@material-ui/icons/Mail';

describe('`ActionButton`', () => {
  let props = {};
  let mountedBtn;
  const shallow = createShallow();

  function btn() {
    if (!mountedBtn) mountedBtn = shallow(<ActionButton {...props} />);
    return mountedBtn;
  }

  beforeEach(() => {
    props = {
      classes: {
        trigger: '',
        actions: ''
      },
      pos: {
        bottom: 0,
        right: 0
      },
      icon: 'message',
      onClick: jest.fn()
    };
    mountedBtn = undefined;
  });

  describe('when `icon` is `message`', () => {
    it('should have a `.trigger` class', () => {
      expect(btn().dive().hasClass('ActionButton-trigger-1')).toBe(true);
    });

    it('should render `Message`', () => {
      expect(btn().dive().find(Message).length).toBe(1);
    });
  });

  describe('when `icon` is `mail`', () => {
    beforeEach(() => {
      btn().setProps({ icon: 'mail' });
    });

    it('should have an `.actions` class', () => {
      expect(btn().dive().hasClass('ActionButton-actions-2')).toBe(true);
    });

    it('should render `Mail`', () => {
      expect(btn().dive().find(Mail).length).toBe(1);
    });
  });

  describe('when `icon` is `linkedin`', () => {
    beforeEach(() => {
      btn().setProps({ icon: 'linkedin' });
    });

    it('should have an `.actions` class', () => {
      expect(btn().dive().hasClass('ActionButton-actions-2')).toBe(true);
    });

    it('should render `Work`', () => {
      expect(btn().dive().find(Work).length).toBe(1);
    });
  });

  describe('when `icon` is `github`', () => {
    beforeEach(() => {
      btn().setProps({ icon: 'github' });
    });

    it('should have an `.actions` class', () => {
      expect(btn().dive().hasClass('ActionButton-actions-2')).toBe(true);
    });

    it('should render `Code`', () => {
      expect(btn().dive().find(Code).length).toBe(1);
    });
  });

  describe('when `icon` is `unknown`', () => {
    beforeEach(() => {
      btn().setProps({ icon: 'unknown' });
    });

    it('should have an `.actions` class', () => {
      expect(btn().dive().hasClass('ActionButton-actions-2')).toBe(true);
    });

    it('should render `Message`', () => {
      expect(btn().dive().find(Message).length).toBe(1);
    });
  });

  it('should invoke the `onClick` prop when clicked', () => {
    btn().simulate('click');
    expect(btn().props().onClick.mock.calls.length).toBe(1);
  });

  it('should match snapshot', () => {
    expect(btn()).toMatchSnapshot();
  });
});
