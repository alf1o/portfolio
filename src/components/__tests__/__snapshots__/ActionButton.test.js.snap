// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`ActionButton\` should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(ActionButton)
    classes={
        Object {
            "actions": "",
            "trigger": "",
          }
    }
    icon="message"
    onClick={[Function]}
    pos={
        Object {
            "bottom": 0,
            "right": 0,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "classes": Object {
        "actions": "ActionButton-actions-2",
        "trigger": "ActionButton-trigger-1",
      },
      "icon": "message",
      "onClick": [Function],
      "pos": Object {
        "bottom": 0,
        "right": 0,
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "classes": Object {
          "actions": "ActionButton-actions-2",
          "trigger": "ActionButton-trigger-1",
        },
        "icon": "message",
        "onClick": [Function],
        "pos": Object {
          "bottom": 0,
          "right": 0,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
